![Alt text: Logo da Zoop](https://github.com/alura-cursos/python-analise-chatgpt-assistente/blob/main/Logos%20e%20paleta%20-%20ZOOP/Logo%20fundo%20colorido%20(2).png?raw=true)

# Aula 1 - Acesso aos dados

In [ ]:
url = 'https://github.com/alura-cursos/python-analise-chatgpt-assistente/raw/main/Dados/dados_vendas.json'

### Prompt:

Vamos atuar como cientistas de dados em uma empresa de supermercados.

Recebemos dados no formato de arquivo JSON, estruturados em listas e dicionários. O arquivo está disponibilizado em uma URL do Github.

Precisamos analisar esses dados para ajudar nas decisões da empresa. Programamos em Python e conhecemos as bibliotecas de análise de dados Numpy e Pandas. Portanto, priorize o uso dessas bibliotecas.

Por favor, nos informe como podemos fazer a leitura desses dados no formato JSON e como podemos visualizá-los em formato de tabela no Python.

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_json(url)
df.head()

item_identificador loja_identificador  vendas_totais  \
0              FDB08             OUT018      176503.58   
1              DRQ35             OUT049      185758.20   
2              FDD14             OUT018      165983.94   
3              FDY37             OUT045      314923.40   
4              FDY59             OUT018       64782.34   

                                                item  \
0  {'item_peso': 6.055, 'item_conteudo_gordura': ...   
1  {'item_peso': 9.3, 'item_conteudo_gordura': 'B...   
2  {'item_peso': 20.7, 'item_conteudo_gordura': '...   
3  {'item_peso': 17.0, 'item_conteudo_gordura': '...   
4  {'item_peso': 8.195, 'item_conteudo_gordura': ...   

                                                loja  
0  {'loja_ano_estabelecimento': 2019, 'loja_taman...  
1  {'loja_ano_estabelecimento': 2009, 'loja_taman...  
2  {'loja_ano_estabelecimento': 2019, 'loja_taman...  
3  {'loja_ano_estabelecimento': 2012, 'loja_taman...  
4  {'loja_ano_estabelecimento': 2019, 'loja_taman...

### Prompt:

O DataFrame "df" possui as colunas "item_identificador", "loja_identificador", "vendas_totais", "item" e "loja".

As colunas "item" e "loja" contêm dicionários aninhados dentro de cada uma das linhas. Como posso transformar os dados dos dicionários e colunas para torná-los mais acessíveis no meu DataFrame?

In [ ]:
# Importe a função json_normalize
from pandas import json_normalize

# Normalize os dados da coluna "item"
df_item_normalized = json_normalize(df['item'])

# Normalize os dados da coluna "loja"
df_loja_normalized = json_normalize(df['loja'])

# Combine os DataFrames normalizados com o DataFrame original
df = pd.concat([df, df_item_normalized, df_loja_normalized], axis=1)

# Exclua as colunas originais de "item" e "loja" se necessário
df.drop(['item', 'loja'], axis=1, inplace=True)
df

item_identificador loja_identificador  vendas_totais  item_peso  \
0                 FDB08             OUT018      176503.58      6.055   
1                 DRQ35             OUT049      185758.20      9.300   
2                 FDD14             OUT018      165983.94     20.700   
3                 FDY37             OUT045      314923.40     17.000   
4                 FDY59             OUT018       64782.34      8.195   
...                 ...                ...            ...        ...   
8545              FDY08             OUT010       28096.76      9.395   
8546              FDC41             OUT017      130163.90     15.600   
8547              NCQ53             OUT045      614533.40     17.600   
8548              FDL46             OUT017      164985.24     20.350   
8549              NCN30             OUT046       96541.00     16.350   

      item_conteudo_gordura  item_visibilidade             item_tipo  \
0     Baixo Teor de Gordura           0.031230     Frutas e Vegetais   
1     Baixo Teor de Gordura           0.042357    Bebidas Alcoólicas   
2     Baixo Teor de Gordura           0.170500             Enlatados   
3                   Regular           0.026623             Enlatados   
4     Baixo Teor de Gordura           0.000000           Confeitaria   
...                     ...                ...                   ...   
8545                Regular           0.286345     Frutas e Vegetais   
8546  Baixo Teor de Gordura           0.117575  Alimentos Congelados   
8547  Baixo Teor de Gordura           0.018944             Mercearia   
8548  baixo teor de gordura           0.054363               Lanches   
8549                    BTG           0.016993               Cereais   

      item_preco item_quantidade_venda  loja_ano_estabelecimento loja_tamanho  \
0         160.36                  None                      2019        Médio   
1         123.24                  None                      2009        Médio   
2         184.13                  None                      2019        Médio   
3         144.25                  None                      2012         None   
4          93.15                  None                      2019        Médio   
...          ...                   ...                       ...          ...   
8545      139.18                  None                      2008         None   
8546       75.67                  None                      2017         None   
8547      237.36                  None                      2012         None   
8548      117.95                  None                      2017         None   
8549       95.74                  None                      2007      Pequeno   

     loja_tipo_localizacao            loja_tipo  
0                  Nível 3  Supermercado Tipo 2  
1                  Nível 1  Supermercado Tipo 1  
2                  Nível 3  Supermercado Tipo 2  
3                  Nível 2  Supermercado Tipo 1  
4                  Nível 3  Supermercado Tipo 2  
...                    ...                  ...  
8545               Nível 3              Mercado  
8546               Nível 2  Supermercado Tipo 1  
8547               Nível 2  Supermercado Tipo 1  
8548               Nível 2  Supermercado Tipo 1  
8549               Nível 1  Supermercado Tipo 1  

[8550 rows x 13 columns]

# Aula 2 - Exploração dos dados

### Prompt:

O DataFrame "df" contém dados referentes às vendas de diversos produtos em várias lojas do supermercado. São fornecidos IDs de produtos e IDs de lojas que comercializaram esses produtos, o total arrecadado em vendas, bem como descrições sobre o produto vendido e a loja que o vendeu.

Estamos iniciando a etapa de exploração de dados, na qual desejamos conhecer os valores informados e identificar se existem valores ausentes, duplicados e outras inconsistências.

Por favor, construa um código para executar a etapa de exploração de dados usando o Python.

Não é necessário utilizar "print()" para mostrar o resultado das consultas.

In [ ]:
# Verificar as primeiras 5 linhas do DataFrame
df.head()

item_identificador loja_identificador  vendas_totais  item_peso  \
0              FDB08             OUT018      176503.58      6.055   
1              DRQ35             OUT049      185758.20      9.300   
2              FDD14             OUT018      165983.94     20.700   
3              FDY37             OUT045      314923.40     17.000   
4              FDY59             OUT018       64782.34      8.195   

   item_conteudo_gordura  item_visibilidade           item_tipo  item_preco  \
0  Baixo Teor de Gordura           0.031230   Frutas e Vegetais      160.36   
1  Baixo Teor de Gordura           0.042357  Bebidas Alcoólicas      123.24   
2  Baixo Teor de Gordura           0.170500           Enlatados      184.13   
3                Regular           0.026623           Enlatados      144.25   
4  Baixo Teor de Gordura           0.000000         Confeitaria       93.15   

  item_quantidade_venda  loja_ano_estabelecimento loja_tamanho  \
0                  None                      2019        Médio   
1                  None                      2009        Médio   
2                  None                      2019        Médio   
3                  None                      2012         None   
4                  None                      2019        Médio   

  loja_tipo_localizacao            loja_tipo  
0               Nível 3  Supermercado Tipo 2  
1               Nível 1  Supermercado Tipo 1  
2               Nível 3  Supermercado Tipo 2  
3               Nível 2  Supermercado Tipo 1  
4               Nível 3  Supermercado Tipo 2

In [ ]:
# Resumo estatístico das colunas numéricas
df.describe()

vendas_totais    item_peso  item_visibilidade   item_preco  \
count   8.550000e+03  7081.000000        8550.000000  8550.000000   
mean    2.181949e+05    12.855023           0.066150   141.007453   
std     1.708098e+05     4.643508           0.051578    62.333062   
min     3.329000e+03     4.555000           0.000000    31.290000   
25%     8.349132e+04     8.770000           0.027024    93.787500   
50%     1.794331e+05    12.600000           0.053978   142.935000   
75%     3.100963e+05    16.850000           0.094646   185.760000   
max     1.308696e+06    21.350000           0.328391   266.890000   

       loja_ano_estabelecimento  
count               8550.000000  
mean                2007.830409  
std                    8.372541  
min                 1995.000000  
25%                 1997.000000  
50%                 2009.000000  
75%                 2014.000000  
max                 2019.000000

In [ ]:
# Informações gerais sobre o DataFrame, incluindo tipos de dados e contagem de valores não nulos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8550 entries, 0 to 8549
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   item_identificador        8550 non-null   object 
 1   loja_identificador        8550 non-null   object 
 2   vendas_totais             8550 non-null   float64
 3   item_peso                 7081 non-null   float64
 4   item_conteudo_gordura     8550 non-null   object 
 5   item_visibilidade         8550 non-null   float64
 6   item_tipo                 8550 non-null   object 
 7   item_preco                8550 non-null   float64
 8   item_quantidade_venda     0 non-null      object 
 9   loja_ano_estabelecimento  8550 non-null   int64  
 10  loja_tamanho              6133 non-null   object 
 11  loja_tipo_localizacao     8550 non-null   object 
 12  loja_tipo                 8550 non-null   object 
dtypes: float64(4), int64(1), object(8)
memory usage: 868.5+ KB


In [ ]:
# Verificar a contagem de valores únicos em cada coluna
df.nunique()

item_identificador          1559
loja_identificador            10
vendas_totais               3493
item_peso                    415
item_conteudo_gordura          5
item_visibilidade           7880
item_tipo                     16
item_preco                  4839
item_quantidade_venda          0
loja_ano_estabelecimento       9
loja_tamanho                   3
loja_tipo_localizacao          3
loja_tipo                      4
dtype: int64

In [ ]:
# Verificar se há valores duplicados no DataFrame
df.duplicated().sum()

27

In [ ]:
# Verificar se há valores ausentes (NaN) em cada coluna
df.isnull().sum()

item_identificador             0
loja_identificador             0
vendas_totais                  0
item_peso                   1469
item_conteudo_gordura          0
item_visibilidade              0
item_tipo                      0
item_preco                     0
item_quantidade_venda       8550
loja_ano_estabelecimento       0
loja_tamanho                2417
loja_tipo_localizacao          0
loja_tipo                      0
dtype: int64

In [ ]:
df['item_conteudo_gordura'].unique()

array(['Baixo Teor de Gordura', 'Regular', 'BTG', 'reg',
       'baixo teor de gordura'], dtype=object)

### Prompt:

O DataFrame "df" contém dados relacionados às vendas durante um trimestre de um supermercado. Desejo utilizá-lo para verificar o desempenho dessas vendas.

Durante a exploração do "df", identificamos algumas inconsistências, como:

- Linhas duplicadas,
- Valores nulos e
- Dados categóricos que foram escritos de maneira não padronizada, dando a ideia de que existem mais categorias do que as que realmente existem.

O que devo fazer para garantir que essas inconsistências que encontrei não afetem minha análise?

# Aula 3 - Limpeza e tratamento

### Prompt:

Durante a exploração do DataFrame "df", identificamos 27 linhas duplicadas usando "df.duplicated().sum()".

O objetivo é analisar os dados em "df" para medir o desempenho de uma empresa durante um trimestre de vendas.

O que devemos fazer com os dados duplicados para garantir que não haja impacto negativo na análise?

In [ ]:
df.drop_duplicates(inplace=True)
df

item_identificador loja_identificador  vendas_totais  item_peso  \
0                 FDB08             OUT018      176503.58      6.055   
1                 DRQ35             OUT049      185758.20      9.300   
2                 FDD14             OUT018      165983.94     20.700   
3                 FDY37             OUT045      314923.40     17.000   
4                 FDY59             OUT018       64782.34      8.195   
...                 ...                ...            ...        ...   
8545              FDY08             OUT010       28096.76      9.395   
8546              FDC41             OUT017      130163.90     15.600   
8547              NCQ53             OUT045      614533.40     17.600   
8548              FDL46             OUT017      164985.24     20.350   
8549              NCN30             OUT046       96541.00     16.350   

      item_conteudo_gordura  item_visibilidade             item_tipo  \
0     Baixo Teor de Gordura           0.031230     Frutas e Vegetais   
1     Baixo Teor de Gordura           0.042357    Bebidas Alcoólicas   
2     Baixo Teor de Gordura           0.170500             Enlatados   
3                   Regular           0.026623             Enlatados   
4     Baixo Teor de Gordura           0.000000           Confeitaria   
...                     ...                ...                   ...   
8545                Regular           0.286345     Frutas e Vegetais   
8546  Baixo Teor de Gordura           0.117575  Alimentos Congelados   
8547  Baixo Teor de Gordura           0.018944             Mercearia   
8548  baixo teor de gordura           0.054363               Lanches   
8549                    BTG           0.016993               Cereais   

      item_preco item_quantidade_venda  loja_ano_estabelecimento loja_tamanho  \
0         160.36                  None                      2019        Médio   
1         123.24                  None                      2009        Médio   
2         184.13                  None                      2019        Médio   
3         144.25                  None                      2012         None   
4          93.15                  None                      2019        Médio   
...          ...                   ...                       ...          ...   
8545      139.18                  None                      2008         None   
8546       75.67                  None                      2017         None   
8547      237.36                  None                      2012         None   
8548      117.95                  None                      2017         None   
8549       95.74                  None                      2007      Pequeno   

     loja_tipo_localizacao            loja_tipo  
0                  Nível 3  Supermercado Tipo 2  
1                  Nível 1  Supermercado Tipo 1  
2                  Nível 3  Supermercado Tipo 2  
3                  Nível 2  Supermercado Tipo 1  
4                  Nível 3  Supermercado Tipo 2  
...                    ...                  ...  
8545               Nível 3              Mercado  
8546               Nível 2  Supermercado Tipo 1  
8547               Nível 2  Supermercado Tipo 1  
8548               Nível 2  Supermercado Tipo 1  
8549               Nível 1  Supermercado Tipo 1  

[8523 rows x 13 columns]

### Prompt:

Identificamos três colunas em "df" que apresentam valores nulos:

- A coluna "item_quantidade_venda" possui 8550 linhas com valores nulos.
- A coluna "item_peso" possui 1469 linhas com valores nulos.
- A coluna "loja_tamanho" possui 2417 linhas com valores nulos.

Houve um grande número de vendas de itens durante o trimestre, e não há justificativa para a ausência de dados na coluna "item_quantidade_venda", o que a torna uma coluna dispensável.

A coluna "item_peso" contém valores do tipo float, representando o peso em gramas dos produtos.

A coluna "loja_tamanho" possui três categorias de tamanho para supermercados.

Como devo lidar com esses dados nulos para garantir que minha análise não seja prejudicada?

In [ ]:
df.drop(columns=['item_quantidade_venda'], inplace=True)


In [ ]:
df['item_peso'].fillna(df['item_peso'].mean(), inplace=True)

In [ ]:
df['loja_tamanho'].fillna('Não registrado', inplace=True)


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8523 entries, 0 to 8549
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   item_identificador        8523 non-null   object 
 1   loja_identificador        8523 non-null   object 
 2   vendas_totais             8523 non-null   float64
 3   item_peso                 8523 non-null   float64
 4   item_conteudo_gordura     8523 non-null   object 
 5   item_visibilidade         8523 non-null   float64
 6   item_tipo                 8523 non-null   object 
 7   item_preco                8523 non-null   float64
 8   loja_ano_estabelecimento  8523 non-null   int64  
 9   loja_tamanho              8523 non-null   object 
 10  loja_tipo_localizacao     8523 non-null   object 
 11  loja_tipo                 8523 non-null   object 
dtypes: float64(4), int64(1), object(7)
memory usage: 865.6+ KB


### Prompt:

A coluna "item_conteudo_gordura" do DataFrame "df" apresenta 5 valores únicos dados por: 'Baixo Teor de Gordura', 'Regular', 'BTG', 'reg' e 'baixo teor de gordura'.

Os valores únicos da coluna "item_conteudo_gordura" deveriam ser apenas 'Baixo Teor de Gordura' e 'Regular', mas foram inseridos sem padronização, o que resultou em mais categorias do que o desejado.

Como posso ajustar os dados da coluna "item_conteudo_gordura" para manter apenas as duas categorias padrões?

In [ ]:
# Mapeie os valores não padronizados para os valores padronizados
mapeamento = {
    'BTG': 'Baixo Teor de Gordura',
    'reg': 'Regular',
    'baixo teor de gordura': 'Baixo Teor de Gordura'
}

# Use a função replace para substituir os valores na coluna
df['item_conteudo_gordura'] = df['item_conteudo_gordura'].replace(mapeamento)

# Verifique os valores únicos após a substituição
df['item_conteudo_gordura'].unique()

array(['Baixo Teor de Gordura', 'Regular'], dtype=object)

# Aula 4 - Visualização dos dados

### Prompt:

Os dados do DataFrame "df" estão limpos e tratados. Agora, desejo iniciar a etapa de construção de gráficos para criar visualizações que ofereçam um bom suporte à análise do desempenho de vendas da empresa.

As colunas disponíveis em "df" são as seguintes:
- "item_identificador": Número único atribuído a cada Item
- "loja_identificador": Nome único da Loja
- "vendas_totais": Valor obtido em vendas totais
- "item_peso": Peso do Item em gramas
- "item_conteudo_gordura": Conteúdo de Gordura do Item
- "item_visibilidade": Valor de posicionamento de cada item.
- "item_tipo": Tipo de utilidade do item
- "item_preço": Preço unitário do item
- "loja_ano_estabelecimento": Ano de Estabelecimento da Loja.
- "loja_tamanho": Tamanho da Loja
- "loja_tipo_localização": Tipo de localização da loja
- "loja_tipo": Tipo de loja.

O objetivo é encontrar padrões ou informações que se destacam quando nos referimos ao total obtido em vendas especificado na coluna "vendas_totais".

Com base nas informações fornecidas, quais tipos de visualizações seriam úteis para atingir o objetivo de analisar o desempenho de vendas?

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "loja_tamanho" com 4 categorias de tamanho para lojas: 'Grande', 'Médio', 'Pequeno' e 'Não registrado'.

Desejo criar uma visualização que permita identificar o total obtido em vendas para as quatro categorias de tamanho de loja e representá-las em um gráfico de barras.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "loja_tamanho".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb" e as barras na cor "#e23155".

Além disso, é necessário exibir o total das vendas em milhões acima de cada coluna, com o texto dessa informação na cor "#19325c" e a estrutura "R$ valor M".

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "loja_tipo" com 4 categorias de tipo para lojas: 'Supermercado Tipo 1', 'Supermercado Tipo 2',  'Mercado', 'Supermercado Tipo 3'.

Desejo criar uma visualização que permita identificar o total obtido em vendas para as quatro categorias de tipo de loja e representá-las em um gráfico de barras.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "loja_tipo".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb" e as barras na cor "#e23155".

Além disso, é necessário exibir o total das vendas em milhões acima de cada coluna, com o texto dessa informação na cor "#19325c" e a estrutura "R$ valor M".

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "item_tipo" com 16 categorias de produtos.

Desejo criar uma visualização que permita identificar o total obtido em vendas por categorias de produtos em um gráfico de barras horizontais.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "item_tipo".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb".

As barras precisam estar na cor "#e23155".

Os valores das barras precisam estar ordenadas para manter um aspecto de funil.

Além disso, é necessário exibir o total das vendas em milhões dentro ou ao lado de cada barra, com o texto dessa informação na cor "#19325c" e na estrutura "R$ valor M".

Não mostre o grid no plot.

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "item_conteudo_gordura" com 2 categorias: 'Baixo Teor de Gordura' e 'Regular'.

Desejo criar uma visualização que permita identificar o total obtido em vendas para as 2 categorias de quantidade de gordura em um gráfico de coluna.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "item_conteudo_gordura".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb".

As colunas de cada categoria devem ter cores específicas: a categoria 'Baixo Teor de Gordura' na cor "#e23155" e a categoria 'Regular' na cor "#acdcd3".

Além disso, é necessário exibir o total das vendas em milhões acima de cada coluna, com o texto dessa informação na cor "#19325c" e na estrutura "R$ valor M".

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.